### PR TITLE
Less cloning in early otherwise branch

### PR DESCRIPTION
**Pull request by simonvandel**
This is probably easiest read commit by commit.

### CODE REVIEW

File: compiler/rustc_mir/src/transform/early_otherwise_branch.rs [hunk 1]
```diff
@@ -41,8 +41,8 @@ impl<'tcx> MirPass<'tcx> for EarlyOtherwiseBranch {
             .flat_map(|(bb_idx, bb)| {
                 let switch = bb.terminator();
                 let helper = Helper { body, tcx };
```
Comment (bugadani) [inline compiler/rustc_mir/src/transform/early_otherwise_branch.rs:43]: Is it necessary to create a different Helper object here in each iteration?

Comment (simonvandel) [reply compiler/rustc_mir/src/transform/early_otherwise_branch.rs]: It's not. Fixed.

File: compiler/rustc_mir/src/transform/early_otherwise_branch.rs [hunk 2]
```diff
@@ -300,7 +299,6 @@ impl<'a, 'tcx> Helper<'a, 'tcx> {
                 // the declaration of the discriminant read. Place of this read is being used in the switch
                 let discr_decl = &self.body.local_decls()[discr_local];
                 let discr_ty = discr_decl.ty;
-                // the otherwise target lies as the last element
```
Comment (jyn514) [inline compiler/rustc_mir/src/transform/early_otherwise_branch.rs:305]: I don't think you meant to remove this comment.

File: compiler/rustc_mir/src/transform/early_otherwise_branch.rs [hunk 3]
```diff
@@ -250,12 +250,6 @@ impl<'a, 'tcx> Helper<'a, 'tcx> {
                 return None;
             }
 
-            // check that the value being matched on is the same. The
-            if this_bb_discr_info.targets_with_values.iter().find(|x| x.0 == value).is_none() {
```
Comment (jyn514) [inline compiler/rustc_mir/src/transform/early_otherwise_branch.rs:257]: Where is this check duplicated? The check above does not look at `targets_with_values` at all.

File: compiler/rustc_mir/src/transform/early_otherwise_branch.rs [hunk 4]
```diff
@@ -270,6 +260,16 @@ impl<'a, 'tcx> Helper<'a, 'tcx> {
...
 
+            // The types of the two adts matched on have to be equal for this optimization to apply
+            if discr_info.type_adt_matched_on != this_bb_discr_info.type_adt_matched_on {
+                trace!(
+                    "NO: types do not match. LHS: {:?}, RHS: {:?}",
+                    discr_info.type_adt_matched_on,
+                    this_bb_discr_info.type_adt_matched_on
+                );
+                return None;
+            }
```
Comment (jyn514) [inline compiler/rustc_mir/src/transform/early_otherwise_branch.rs:269]: There doesn't seem to be much point to this change without benchmarks. Personally I would expect a simple `!=` to be faster than two conditionals.

File: compiler/rustc_mir/src/transform/early_otherwise_branch.rs [hunk 5]
```diff
@@ -198,36 +196,36 @@ struct OptimizationToApply<'tcx> {
 struct OptimizationInfo<'tcx> {
     /// Info about the first switch and discriminant
     first_switch_info: SwitchDiscriminantInfo<'tcx>,
-    /// Info about the second switch and discriminant
-    second_switch_info: SwitchDiscriminantInfo<'tcx>,
+    /// Info about all swtiches that are successors of the first switch
```
Comment (jyn514) [inline compiler/rustc_mir/src/transform/early_otherwise_branch.rs:197]:
```suggestion
    /// Info about all switches that are successors of the first switch
```

**Comment by oli-obk** [comment]
No measurable impact. I'll go over the code later, but I think we should thus only take the commits that improve readability

**Comment by camelid** [comment]
@simonvandel Could you address Oli's comments? By the way, you have some merge conflicts.

**Comment by simonvandel** [comment]
I currently don't have the time to work on this, unfortunately. If you prefer to close it for now, then feel tree to do so.

**Pull request closed without merging**